packages/transform/src/lib/TransformMiddleware.ts

```diff
@@ -0,0 +1,485 @@
+import {
+  ApplyMiddlewareOptions,
+  isEveryByHookPayload,
+  isEveryByValuePayload,
+  isMapByHookPayload,
+  isMapByPathPayload,
+  isSomeByHookPayload,
+  isSomeByValuePayload,
+  JoshMiddleware,
+  Method,
+  Payload,
+  PostProvider,
+  PreProvider,
+  resolveVersion,
+  Semver
+} from '@joshdb/provider';
+import { Awaitable, objectToTuples } from '@sapphire/utilities';
+
+@ApplyMiddlewareOptions({ name: 'transform' })
+export class TransformMiddleware<BeforeValue = unknown, AfterValue = unknown> extends JoshMiddleware<
+  // @ts-expect-error 2322 - 'ContextData<BeforeValue, AfterValue>' has properties in common with type 'Context', ts thinks they are incompatible
+  TransformMiddleware.ContextData<BeforeValue, AfterValue>,
+  AfterValue
+> {
+  public get version(): Semver {
+    return resolveVersion('[VI]{version}[/VI]');
+  }
+
+  @PreProvider()
+  public async [Method.Each]<ReturnValue = BeforeValue>(payload: Payload.Each<ReturnValue>): Promise<Payload.Each<ReturnValue>> {
+    const { after } = this.context;
+    const hook = async (value: AfterValue, key: string) => {
+      return payload.hook((await after(value, key, null)) as ReturnValue, key) as ReturnValue;
+    };
+
+    await this.provider[Method.Each]({ ...payload, hook });
+
+    return payload;
+  }
+
+  @PreProvider()
+  public async [Method.Ensure]<ReturnValue = BeforeValue, StoredValue = AfterValue>(
+    payload: Payload.Ensure<ReturnValue>
+  ): Promise<Payload.Ensure<StoredValue>> {
+    const { key, defaultValue } = payload;
+    const { before, autoTransform } = this.context;
+
+    payload.defaultValue = (await before(defaultValue as unknown as BeforeValue, key, null)) as ReturnValue;
+
+    await this.provider[Method.Ensure](payload as unknown as Payload.Ensure<AfterValue>);
+
+    if (!(await this.isTransformed(key))) {
+      if (autoTransform === true) {
+        const { data } = await this.provider[Method.Get]({ method: Method.Get, errors: [], key, path: [] });
+
+        await this.provider[Method.Set]({ method: Method.Set, errors: [], key, path: [], value: data });
+        await this.updateMetadataPath(key, this.objectPathKeys(data));
+      }
+
+      process.emitWarning(
+        `The ensured data at "${key}" has not been transformed yet, please enable "autoTransform" to transform the data automatically or set() the data at the key to transform it.`
```

Comment:
Should this be an else block for the warning?

Reply:
I believe it was, I am not sure if I pulled it out to see if it would fix the codecov I must have left it outside of it.